aten/src/ATen/native/TensorShape.cpp

```diff
@@ -564,11 +616,25 @@ TORCH_IMPL_FUNC(cat_out_cpu)
 
   auto materialized = tensors.materialize();
 
-  // fast path for single thread when both inputs and result are contiguous and not empty
   bool use_serial_kernel = result.numel() < at::internal::GRAIN_SIZE || at::get_num_threads() == 1;
   ScalarType dtype = materialized[valid].get().scalar_type();
-  bool serial_dtype = at::isFloatingType(dtype);
-  if (use_serial_kernel && all_contiguous && all_same_dtype && serial_dtype) {
+  bool serial_dtype_float = at::isFloatingType(dtype);
+  bool serial_dtype_int = at::isIntegralType(dtype, /*includeBool=*/true);
+  // fast path for single thread when both inputs and result are contiguous and
+  // not empty, and concat dim is 0 or 1
+  if (use_serial_kernel && all_contiguous && all_same_dtype &&
+      (MemoryFormat::Contiguous == memory_format) && (serial_dtype_int || serial_dtype_float)) {
```

Comment:
Actually, why do you need to know the dtype at all? lol

Reply:
Just to double check the complex types, their storage are also contiguous, so safe to use memcpy to copy over?